machine_test.go

```diff
@@ -1081,37 +1081,92 @@ func TestCaptureFifoToFile_leak(t *testing.T) {
 	assert.Contains(t, loggerBuffer.String(), `file already closed`, "log")
 }
 
-func TestWaitWithKill(t *testing.T) {
+// Replace filesystem-unsafe characters (such as /) which are often seen in Go's test names
+var fsSafeTestName = strings.NewReplacer("/", "_")
+
+func TestWait(t *testing.T) {
 	fctesting.RequiresRoot(t)
-	ctx := context.Background()
 
-	socketPath := filepath.Join(testDataPath, t.Name())
-	defer os.Remove(socketPath)
+	cases := []struct {
+		name string
+		stop func(m *Machine)
+	}{
+		{
+			name: "StopVMM",
+			stop: func(m *Machine) {
+				err := m.StopVMM()
+				require.NoError(t, err)
+			},
+		},
+		{
+			name: "Kill",
+			stop: func(m *Machine) {
+				pid, err := m.PID()
+				require.NoError(t, err)
+
+				process, err := os.FindProcess(pid)
+				err = process.Kill()
+				require.NoError(t, err)
+			},
+		},
+		{
+			name: "Context Cancel",
+		},
+	}
 
-	cfg := createValidConfig(t, socketPath)
-	cmd := VMCommandBuilder{}.
-		WithSocketPath(cfg.SocketPath).
-		WithBin(getFirecrackerBinaryPath()).
-		Build(ctx)
-	m, err := NewMachine(ctx, cfg, WithProcessRunner(cmd))
-	require.NoError(t, err)
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			ctx := context.Background()
+			vmContext, vmCancel := context.WithCancel(context.Background())
 
-	err = m.Start(ctx)
-	require.NoError(t, err)
+			socketPath := filepath.Join(testDataPath, fsSafeTestName.Replace(t.Name()))
+			defer os.Remove(socketPath)
 
-	go func() {
-		pid, err := m.PID()
-		require.NoError(t, err)
+			cfg := createValidConfig(t, socketPath)
+			cmd := VMCommandBuilder{}.
+				WithSocketPath(cfg.SocketPath).
+				WithBin(getFirecrackerBinaryPath()).
+				Build(vmContext)
+			m, err := NewMachine(ctx, cfg, WithProcessRunner(cmd))
+			require.NoError(t, err)
 
-		process, err := os.FindProcess(pid)
-		require.NoError(t, err)
+			err = m.Start(ctx)
+			require.NoError(t, err)
 
-		err = process.Kill()
-		require.NoError(t, err)
-	}()
+			pid, err := m.PID()
+			require.NoError(t, err)
 
-	err = m.Wait(ctx)
-	require.Error(t, err, "Firecracker was killed and it must be reported")
+			go func() {
+				if c.stop != nil {
+					c.stop(m)
+				} else {
+					vmCancel()
+				}
+			}()
+
+			err = m.Wait(ctx)
+			require.Error(t, err, "Firecracker was killed and it must be reported")
+
+			alive, err := isProcessAlive(pid)
+			require.False(t, alive, "pid=%d is still there", pid)
+		})
+	}
+}
+
+func isProcessAlive(pid int) (bool, error) {
+	// Using kill(2) with signal=0 to check the existence of the process,
+	// because os.FindProcess always returns a process, regardless of whether the process is
+	// alive or not.
```

Comment:
Can you elaborate on this? 
As far as I see go's FindProcess does the same:
https://golang.org/src/os/exec_unix.go#L88
https://golang.org/src/os/exec_unix.go#L73
and its expected to receive err `os: process already finished`

Reply:
Clarification: you can use it like `err := os.FindProcess(x).Signal(0)`

Reply:
The error from Process#Signal() is not typed. The only way to check the type of the error is string comparison. So I'm unsure that is reliable over time.

https://github.com/golang/go/blob/753d56d3642eb83848aa39e65982a9fc77e722d7/src/os/exec_unix.go#L53

Reply:
Don't think you need to cast the error
https://github.com/golang/go/blob/753d56d3642eb83848aa39e65982a9fc77e722d7/src/os/exec_unix.go#L72-L74

Reply:
kill(2) returns EINVAL, EPERM or ESRCH as an error, but Go's Signal() doesn't distinguish them as different errors.

Would it be better to do something like below? Personally I don't want to depend on the string representation because it might change.

```
if err.Error() == "os: process already finished" {
  ...
}
```

Thoughts?

Reply:
According to `Signal` it will only return `os: process already finished` error when `syscall.ESRCH` occurs. So, to me I think it doesn't matter either way. I think the string comparison is more ugly than checking against `ESRCH`, but I am fine with either solution.

Reply:
> Apparently it sets done based on Wait, which may return the already finished error, which complicates the checks.

And all of these are happening inside os.Process. syscall.Kill is much simpler I think.

Reply:
I'd use os.Process, but I agree that strings comparison is ugly. Unfortunately sometimes its hard to get away from it (for instance there is a similar checker in containerd: https://github.com/containerd/containerd/blob/b0821c801dc2225bb7478f91e967888a353fb60a/pkg/process/utils.go#L131)

Reply:
Okay. The last revision uses os.Process instead.